site/_posts/2018-02-12-discovering-a-ruby-memory-leak.html.md

```diff
@@ -0,0 +1,441 @@
+5. fetch queue work to assign to worker processes
+
+Each of these steps happen in varying intervals, but mostly happen on a
+consistent basis.  Also if everything is running "normally", the work that is
+done is also fairly consistent and repeatable across each one of those steps.
+
+After trying to instrument the main loop with tools like `MemoryProfiler`,
+`stackprof`, and even just looking at regular ruby memory dumps using
+`ObjectSpace.dump_all`, we ended up taking a more low tech approach of just
+removing code by commenting out the steps above, and seeing if the leak still
+happened after a server restart.
+
+To start off, we already had a baseline of what it looked like with the code
+uncommented, and it leaked pretty consistently, so we did the inverse and
+commented out all of the steps above to confirm the leak stopped (it did,
+thankfully).  We then worked our way back to the baseline by re-adding the
+methods one by one, and observing if the leak returned.  After doing that a few
+times, we eventually determined that uncommenting the
+`MiqServer#monitor_workers` code (`4.` from the list above) was the cause.
+Progress!
+
+Now, `MiqServer#monitor_workers` is [pretty dense][8], so this wasn't the end
+of the line, but it allowed us to start drilling in further and try some other
+tests outside the context of the application.
+
+
+Step 3:  First Observations
+---------------------------
+
+The tricks from `1.5` were applied at this point, and we started making some
+more nuanced observations about what affected the leak:
+
+* The server seemed to leak faster with more workers forked and monitored by
+  the server process. This meant to us that it was probably calling a method
+  that was dependent on the number workers.
+* While adding some instrumentation to do some regular `GC.start`s and logging
+  of `GC.stat`, we noticed the number of ruby objects didn't really increase
+  (some custom caching that we do caused us to question this for a bit).
+* The `[heap]` in `/proc/<PID>/smaps` was the only thing growing.  Combined
+  with the info from the last bullet, this meant that the leak surely had to be
+  in a c-extension, or ruby itself.  (an important observation made by Joe
+  Rafaniello)
+* None of the suspected code looked to be a possible cause of a leak so it was
+  probably something specific to our environment.
+
+Those were all important observations, but the last one was a double-edged
+sword.  Throughout this process, we continued to hold the mindset that "this
+has to be something with our code", and it led us to going down a lot of wrong
+rabbit holes searching for answers that re-enforced that bias.  Normally it is
+an accurate and safe bias to have, but as mentioned at the beginning, we were
+part of that small case where it wasn't.
+
+
+Step 4:  Lots of failure, torment, and misery
+---------------------------------------------
+
+This is when we started going outside of the ManageIQ application and started
```

Comment:
"We eventually started going outside..."

Reply:
Also s/going/looking/

Reply:
Edit can be found here:

https://gist.github.com/NickLaMuro/730b252a48019cc6bef85270ac16ff44/revisions#diff-4756561d24941a282634640442f5d8be

Commit message comments:

I liked the suggestion to use "looking" instead of "going".  Feels less clunky.  I also switched s/application/code base/ since I think that is  more accurate, and again, less clunky.
    
I didn't, however, use "we eventually" since that is used in 3 other places in the post.  I changed it to open with "At this point", which seemed less clunky.